### ISSUE TITLE
Not Working on Apple Silicon MacOS.

### ISSUE THREAD

**Issue by sgeos** [daniel5151/libc_alloc#7]
## Bug Report

`libc_alloc` is not working out of the box on MacOS. So far as I can tell, MacOS has `memalign_posix()`, but not `memalign()`.

## Software Versions

```bash
$ date -u "+%Y-%m-%d %H:%M:%S +0000"
2024-03-16 21:19:01 +0000
$ uname -vm
Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000 arm64
$ ex -s +'%s/<[^>].\{-}>//ge' +'%s/\s\+//e' +'%norm J' +'g/^$/d' +%p +q! /System/Library/CoreServices/SystemVersion.plist | grep -E 'ProductName|ProductVersion' | sed 's/^[^ ]* //g' | sed 'N; s/\n/ /g'
macOS 14.2.1
$ sysctl -n machdep.cpu.brand_string
Apple M1 Max
$ cargo --version
cargo 1.78.0-nightly (7065f0ef4 2024-03-12)
```

## Failure

```bash
$ cargo run
* snip *
  = note: ld: Undefined symbols:
            _memalign, referenced from:
                _$LT$libc_alloc..LibcAlloc$u20$as$u20$core..alloc..global..GlobalAlloc$GT$::alloc::h3c9bf5ea6ef1efb3 in flight_brain-e59ec13c5f5719a4.53s8p6zd7cw3ughd.rcgu.o
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
* snip *
```

## Local Fix

Adding an exposed `memalign()` function that delegates to `memalign_posix()` resulted in a successful build, but it seems like any sort of conditional compilation should be baked into `libc_alloc`.

```rust
#[cfg(target_os = "macos")]
#[no_mangle]
pub unsafe extern "C" fn memalign(align: size_t, size: size_t) -> *mut c_void {
    let mut memptr: *mut c_void = core::ptr::null_mut();
    let result = libc::posix_memalign(&mut memptr, align, size);
    if result == 0 { memptr } else { core::ptr::null_mut() }
}
```

**Comment by daniel5151** [comment]
Interestingly enough, I just-so-happened to see that someone actually forked libc_alloc to get it working on macOS:
https://github.com/DmitryDodzin/libc_alloc/tree/posix

@DmitryDodzin, not sure if you were planning on sending in a PR with those changes, but I'd be happy if you did!

Alternatively, feel free to take-point on sending a PR in yourself @sgeos.

Either way, I'd be happy to review any PR that fixes libc_alloc on macOS, and would be able to publish a new version of the crate pretty quickly.

**Comment by sgeos** [comment]
PR sent. I reviewed @DmitryDodzin's branch and could not figure out what he was doing with errno, so I went with the submitted approach.

**Comment by daniel5151** [comment]
Indeed - I'm also not entirely sure what the errno stuff was about...

Well, I never said I actually reviewed that code - just that I saw that the fork was made, heh 😅

In any case, thanks for sending in the PR. Once it lands, we should be able to mark this issue as resolved.

**Comment by sgeos** [comment]
As far as testing goes, all I can say is that:
A) it is working for me on my local machine, and
B) the code looks correct to me.

Is there a project I can run to properly test that everything works as intended on MacOS?

If I have any concerns, it would be that I swear libc_alloc used to work on MacOS.

**Comment by DmitryDodzin** [comment]
Yeah, I was planning on opening a PR from the posix branch though I didn't yet because I was unsure about if Im handling errno correctly and didn't quite had the time (a bit of skill issue as well) to test the errors cases, my entire concern stems from [posix_memalign(3)](https://man7.org/linux/man-pages/man3/posix_memalign.3.html) and more specifically the RETURN_VALUE where I see that `memalign` does set errno

> posix_memalign() returns zero on success, or one of the error
       values listed in the next section on failure.  The value of [errno](https://man7.org/linux/man-pages/man3/errno.3.html)
       is not set.

**Comment by sgeos** [comment]
> Yeah, I was planning on opening a PR from the posix branch though I didn't yet because I was unsure about if Im handling errno correctly

Is there any reason to set errno? Does your code actually work to set errno? My reading is "posix_memalign returns errors directly, so do not worry about errno".

Also, thank you for the branch. It was useful to have something to review.

**Comment by daniel5151** [comment]
Closed via #8

**Comment by daniel5151** [comment]
libc_alloc 1.0.7 is now published to crates.io with @sgeos's PR.

Thanks again for sending it in!
